Formset test failures on Django 1.11a1

Several tests are failing against Django 1.11a1, due to this change:

https://github.com/django/django/commit/181f492ad021aeb43105aa9d38106ad7baf00211
https://code.djangoproject.com/ticket/27416

	Creating test database for alias 'default'...
	System check identified no issues (0 silenced).
	..............................FF.........F..........F........................x..
	======================================================================
	FAIL: test_formsets_from_model_superclass_are_exposed (tests.tests.test_cluster_form.ClusterFormTest)
	----------------------------------------------------------------------
	Traceback (most recent call last):
	  File "/home/vagrant/libs/django-modelcluster/tests/tests/test_cluster_form.py", line 451, in test_formsets_from_model_superclass_are_exposed
	    self.assertEqual(instance.reviews.count(), 1)
	AssertionError: 0 != 1

	======================================================================
	FAIL: test_formsets_from_model_superclass_with_explicit_formsets_def (tests.tests.test_cluster_form.ClusterFormTest)
	----------------------------------------------------------------------
	Traceback (most recent call last):
	  File "/home/vagrant/libs/django-modelcluster/tests/tests/test_cluster_form.py", line 482, in test_formsets_from_model_superclass_with_explicit_formsets_def
	    self.assertEqual(instance.reviews.count(), 1)
	AssertionError: 0 != 1

	======================================================================
	FAIL: test_child_updates_without_ids (tests.tests.test_formset.ChildFormsetTest)
	----------------------------------------------------------------------
	Traceback (most recent call last):
	  File "/home/vagrant/libs/django-modelcluster/tests/tests/test_formset.py", line 240, in test_child_updates_without_ids
	    self.assertEqual(2, beatles.members.count())
	AssertionError: 2 != 1

	======================================================================
	FAIL: test_incoming_formset_data (tests.tests.test_formset.TransientFormsetTest)
	----------------------------------------------------------------------
	Traceback (most recent call last):
	  File "/home/vagrant/libs/django-modelcluster/tests/tests/test_formset.py", line 43, in test_incoming_formset_data
	    self.assertEqual(2, len(members))
	AssertionError: 2 != 1

	----------------------------------------------------------------------
	Ran 80 tests in 0.608s

	FAILED (failures=4, expected failures=1)

--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,8 +1,9 @@
 Changelog
 =========
 
-Unreleased
-~~~~~~~~~~
+3.1 (xx.xx.xxxx)
+~~~~~~~~~~~~~~~~
+* Django 1.11 compatibility
 * Added the ability to install the optional dependency `django-taggit`
   using `pip install django-modelcluster[taggit]`
 

--- a/modelcluster/forms.py
+++ b/modelcluster/forms.py
@@ -1,12 +1,11 @@
 from __future__ import unicode_literals
 
+import django
 from django.utils.six import with_metaclass
-
 from django.forms.models import (
     BaseModelFormSet, modelformset_factory,
     ModelForm, _get_foreign_key, ModelFormMetaclass, ModelFormOptions
 )
-
 from django.db.models.fields.related import ForeignObjectRel
 
 
@@ -66,6 +65,16 @@ class BaseTransientModelFormSet(BaseModelFormSet):
                 if not commit:
                     self.saved_forms.append(form)
         return saved_instances
+
+    if django.VERSION < (1, 9):
+        def save_existing(self, form, instance, commit=True):
+            """Saves and returns an existing model instance for the given form."""
+            return form.save(commit=commit)
+
+        def delete_existing(self, obj, commit=True):
+            """Deletes an existing model instance."""
+            if commit:
+                obj.delete()
 
 
 def transientmodelformset_factory(model, formset=BaseTransientModelFormSet, **kwargs):

--- a/modelcluster/forms.py
+++ b/modelcluster/forms.py
@@ -16,6 +16,8 @@ from modelcluster.models import get_all_child_relations
 class BaseTransientModelFormSet(BaseModelFormSet):
     """ A ModelFormSet that doesn't assume that all its initial data instances exist in the db """
     def _construct_form(self, i, **kwargs):
+        # Need to override _construct_form to avoid calling to_python on an empty string PK value
+
         if self.is_bound and i < self.initial_form_count():
             pk_key = "%s-%s" % (self.add_prefix(i), self.model._meta.pk.name)
             pk = self.data[pk_key]
@@ -37,6 +39,33 @@ class BaseTransientModelFormSet(BaseModelFormSet):
 
         # bypass BaseModelFormSet's own _construct_form
         return super(BaseModelFormSet, self)._construct_form(i, **kwargs)
+
+    def save_existing_objects(self, commit=True):
+        # Need to override _construct_form so that it doesn't skip over initial forms whose instance
+        # has a blank PK (which is taken as an indication that the form was constructed with an
+        # instance not present in our queryset)
+
+        self.changed_objects = []
+        self.deleted_objects = []
+        if not self.initial_forms:
+            return []
+
+        saved_instances = []
+        forms_to_delete = self.deleted_forms
+        for form in self.initial_forms:
+            obj = form.instance
+            if form in forms_to_delete:
+                if obj.pk is None:
+                    # no action to be taken to delete an object which isn't in the database
+                    continue
+                self.deleted_objects.append(obj)
+                self.delete_existing(obj, commit=commit)
+            elif form.has_changed():
+                self.changed_objects.append((obj, form.changed_data))
+                saved_instances.append(self.save_existing(form, obj, commit=commit))
+                if not commit:
+                    self.saved_forms.append(form)
+        return saved_instances
 
 
 def transientmodelformset_factory(model, formset=BaseTransientModelFormSet, **kwargs):

